src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/23.diff.sql

```diff
@@ -0,0 +1,559 @@
+	WITH (DATA_COMPRESSION = PAGE, ONLINE=ON) 
+END
+
+GO
+
+-- CompartmentAssignment
+-- We are creating primary key on the same set of columns for which clustered index exists.
+-- If script execution failed immediately after dropping clustered index, then on the same set of columns we have non clustered index for non-historical records, so we should be good.
+EXEC dbo.LogSchemaMigrationProgress 'Dropping IXC_CompartmentAssignment'
+IF EXISTS (
+    SELECT * 
+	FROM sys.indexes 
+	WHERE name='IXC_CompartmentAssignment' AND object_id = OBJECT_ID('dbo.CompartmentAssignment'))
+BEGIN
+	DROP INDEX IXC_CompartmentAssignment ON dbo.CompartmentAssignment
+	WITH (ONLINE=ON)
+END
+
+-- Add clustered primary key on the same set of columns for which clustered index just dropped above.
+EXEC dbo.LogSchemaMigrationProgress 'Adding PKC_CompartmentAssignment'
+IF NOT EXISTS (
+    SELECT * 
+	FROM sys.key_constraints 
+	WHERE name='PKC_CompartmentAssignment' AND type='PK')
+BEGIN
+	ALTER TABLE dbo.CompartmentAssignment 
+	ADD CONSTRAINT PKC_CompartmentAssignment PRIMARY KEY CLUSTERED(ResourceTypeId, ResourceSurrogateId, CompartmentTypeId, ReferenceResourceId)
+	WITH (DATA_COMPRESSION = PAGE, ONLINE=ON) 
+	ON PartitionScheme_ResourceTypeId(ResourceTypeId)
+END
+
+GO
+
+-- Resource
+-- Since we need to create primary key on the same of columns for which clustered index exists.
+-- We will create non-clustered index first, drop IXC, create PKC and finally drop IX
+-- so that the table would never be left without indexes.
+-- Creating temporary non-clustered index
+EXEC dbo.LogSchemaMigrationProgress 'Creating IX_Resource'
+IF NOT EXISTS (
+    SELECT * 
+	FROM sys.indexes 
+	WHERE name='IX_Resource' AND object_id = OBJECT_ID('dbo.Resource'))
+BEGIN
+	CREATE NONCLUSTERED INDEX IX_Resource ON dbo.Resource
+	(
+		ResourceTypeId,
+		ResourceSurrogateId
+	)
+	WITH (DATA_COMPRESSION = PAGE, ONLINE=ON)
+	ON PartitionScheme_ResourceTypeId(ResourceTypeId)
+END
+
+-- Deleting clustered index
+EXEC dbo.LogSchemaMigrationProgress 'Dropping IXC_Resource'
+IF EXISTS (
+    SELECT * 
+	FROM sys.indexes 
+	WHERE name='IXC_Resource' AND object_id = OBJECT_ID('dbo.Resource'))
+BEGIN
+	DROP INDEX IXC_Resource ON dbo.Resource
+	WITH (ONLINE=ON)
+END
+
+-- Adding clustered primary key
+EXEC dbo.LogSchemaMigrationProgress 'Adding PKC_Resource'
+IF NOT EXISTS (
+    SELECT * 
+	FROM sys.key_constraints 
+	WHERE name='PKC_Resource' AND type='PK')
+BEGIN
+	ALTER TABLE dbo.Resource 
+	ADD CONSTRAINT PKC_Resource PRIMARY KEY CLUSTERED(ResourceTypeId, ResourceSurrogateId)
+	WITH (DATA_COMPRESSION = PAGE, ONLINE=ON) 
+	ON PartitionScheme_ResourceTypeId(ResourceTypeId)
+END
+
+-- Deleting temporary non-clustered index created.
+EXEC dbo.LogSchemaMigrationProgress 'Dropping IX_Resource'
+IF EXISTS (
+    SELECT * 
+	FROM sys.indexes 
+	WHERE name='IX_Resource' AND object_id = OBJECT_ID('dbo.Resource'))
+BEGIN
+	DROP INDEX IX_Resource ON dbo.Resource
```

Comment:
Do we need "ONLINE=ON" here?

Reply:
Only a clustered index can be dropped online.